tools/encryption/encryptionutils/encryption-helper.go

```diff
@@ -0,0 +1,243 @@
+    "encoding/base64"
+	"fmt"
+	"io"
+	"golang.org/x/net/html"
+	"golang.org/x/net/html/atom"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/tink/go/aead"
+	"github.com/google/tink/go/hybrid"
+	"github.com/google/tink/go/insecurecleartextkeyset"
+	"github.com/google/tink/go/keyset"
+	"github.com/google/tink/go/core/registry"
+	gcmpb "github.com/google/tink/proto/aes_gcm_go_proto"
+	tinkpb "github.com/google/tink/proto/tink_go_proto"
+	"net/http"
+	"strings"
+)
+
+const AES_GCM_KEY_URL string = "type.googleapis.com/google.crypto.tink.AesGcmKey" 
+const AES_GCM_KEY_SIZE uint32 = 16
+
+// Public function to generate an encrypted HTML document given the original.
+func GenerateEncryptedDocument(html_str string, public_key_url string, access_requirement string) (string, error) {
+	keyManager, err := registry.GetKeyManager(AES_GCM_KEY_URL)
+	if err != nil {
+		return "", err
+	}
+	key, err := generateNewAesGcmKey(keyManager)
+	if err != nil {
+		return "", err
+	}
+	ks := createAesGcmKeyset(key)
+	r := strings.NewReader(html_str)
+	parsed_html, err := html.Parse(r)
+	if err != nil {
+		return "", err
+	}
+	encrypted_sections := getAllEncryptedSections(parsed_html)
+	kh, err := insecurecleartextkeyset.Read(&keyset.MemReaderWriter{Keyset: &ks})
+	if err != nil {
+		return "", err
+	}
+	err = encryptAllSections(parsed_html, encrypted_sections, kh)
+	if err != nil {
+		return "", err
+	}
+	google_public_key, err  := getGooglePublicKey(public_key_url)
+	if err != nil {
+		return "", err
+	}
+	ks_enc, err := proto.Marshal(&ks)
+	if err != nil {
+		return "", err
+	}
+	encrypted_key, err := encryptDocumentKey(base64.StdEncoding.EncodeToString(ks_enc), access_requirement, google_public_key)
+	if err != nil {
+		return "", err
+	}
+	if err := addEncryptedDocumentKeyToHead(encrypted_key, parsed_html); err != nil {
+		return "", err
+	}
+	return renderNode(parsed_html, false), nil
+}
+
+// Generates a new AES-GCM key.
+func generateNewAesGcmKey(km registry.KeyManager) ([]byte, error) {
+	serialized_proto, _ := proto.Marshal(&gcmpb.AesGcmKeyFormat{KeySize: AES_GCM_KEY_SIZE})
+	m, err := km.NewKey(serialized_proto)
+	if err != nil {
+		return nil, err
+	}
+	return proto.Marshal(m)
+}
+
+// Creates an AES-GCM Keyset using the input key.
+func createAesGcmKeyset(key []byte) tinkpb.Keyset {
+	keyData := tinkpb.KeyData{
+		KeyMaterialType: tinkpb.KeyData_SYMMETRIC,
+		TypeUrl: AES_GCM_KEY_URL,
+		Value: key,
+	}
+	keys := []*tinkpb.Keyset_Key{
+		&tinkpb.Keyset_Key{
+			KeyData:          &keyData,
+			Status:           tinkpb.KeyStatusType_ENABLED,
+			KeyId:            1,
+			OutputPrefixType: tinkpb.OutputPrefixType_TINK,
+		},
+	}
+    return tinkpb.Keyset{
+		PrimaryKeyId: 1,
+		Key:          keys,
+	}
+}
+
+// Retrieves all encrypted content sections from the parsed HTML tree.
+func getAllEncryptedSections(parsed_html *html.Node) []*html.Node {
+	var encrypted_sections []*html.Node
+	for n := parsed_html.FirstChild; n != nil; n = n.NextSibling {
```

Comment:
There are a lot of for loops in here. I got a little lost. Is this trying to traverse the entire tree depth first? There are a couple of ways do it. One is recursion. Or there's a simpler way to traverse the tree. I coded it for the amppackager [here](https://github.com/ampproject/amppackager/blob/releases/transformer/internal/htmlnode/htmlnode.go#L47).
Example usage [here](https://github.com/ampproject/amppackager/blob/releases/transformer/transformers/nodecleanup.go#L47).

Reply:
Oh, I see, this is trying to find the <body> and then traverse everything within. Recommend splitting out FindBody from  the processing that's being done.

Reply:
I agree this is a confusing function. I moved the DFS function into it's own function--hopefully that clears things up.

Reply:
I agree it is quite confusing. I moved the DFS algorithm into its own function to clear things up.